[RPC] Split part of validateaddress into getaddressinfo

achow101 opened the pull request:
This PR makes a new RPC command called `getaddressinfo` which relies on the wallet. It contains all of `validateaddress`'s address info stuff. Those parts in `validateaddress` have been marked as deprecated. The tests have been updated to use `getaddressinfo` except the `disablewallet` test which is the only test that actually uses `validateaddress` to validate an address.

achow101 committed:
Create getaddressinfo RPC and deprecate parts of validateaddress

Moves the parts of validateaddress which require the wallet into getaddressinfo
which is part of the wallet RPCs. Mark those parts of validateaddress which
require the wallet as deprecated.

Validateaddress will  call getaddressinfo
for the data that both share for right now.

Moves IsMine functions to libbitcoin_common and then links libbitcoin_wallet
before libbitcoin_common in order to prevent linker errors since IsMine is no
longer used in libbitcoin_server.

laanwj commented:
Concept ACK. Makes a lot of sense to move the wallet-specific information to a separate RPC.

gmaxwell reviewed (commented):
Concept ACK.

sipa commented:
Currently this duplicates a singificant amount of code. Would it be possible to abstract out the 'info' part of `getaddressinfo`, and then call that from `validateaddress` as well (temporarily, until removed)?

achow101 commented:
@sipa Apparently, no, this is not possible. My attempt at doing so has resulted in a linker failure. Apparently there is some makefile problems with that. https://github.com/achow101/bitcoin/tree/getaddressinfo-broken is the branch with the duplication removed, but there is a linker error with `IsMine` calls. The full build error is here: https://gist.github.com/achow101/0f0e2d9e75d9d77132ab83a7605f2377

achow101 commented:
On second thought, I think I fixed the problem.

sipa commented:
I believe you can do:

```diff
 bitcoind_LDADD = \
   $(LIBBITCOIN_SERVER) \
+  $(LIBBITCOIN_WALLET) \
   $(LIBBITCOIN_COMMON) \
   $(LIBUNIVALUE) \
   $(LIBBITCOIN_UTIL) \
-  $(LIBBITCOIN_WALLET) \
   $(LIBBITCOIN_ZMQ) \
```
in `Makefile.am`.

Then the _ismine code can move from wallet to common (which I think belongs there, as signing code is already in common too).

achow101 commented:
@sipa I did that in my latest commit, but it seems that travis doesn't like it.

laanwj commented:
> Then the _ismine code can move from wallet to common (which I think belongs there, as signing code is already in common too).

From what I remember it was already moved in the other direction at some point for the same reason.
The circular dependency between `_wallet` and `_server` remains a problem.
We'll need to solve #7965 to get rid of it.
Unfortunately, this is part of that...

jnewbery commented:
needs rebase

achow101 commented:
rebased

jnewbery commented:
Concept ACK. This PR is currently difficult to review because there are broken intermediate commits, commits which reverse previous commits, random style changes in unconnected commits, etc. Are you able to tidy this up to make it easier to review? If not, I'm happy to take this PR and tidy it up.

My high-level feedback is that we should hide the deprecated fields in `validateaddress` behind a command-line argument (similar to #11031). The `ismine` move to libbitcoin_common seems fine to me.

achow101 commented:
Rebased and tidied up

jnewbery commented:
Travis is failing on invalid scripted diff check.

jnewbery reviewed (commented):
A few nits inline. I think you could drop the third commit by making your scripted diff more focussed:

```
find ./test/functional -path '*py' -not -path ./test/functional/disablewallet.py -exec sed -i'' -e 's/validateaddress/getaddressinfo/g' {} \;
```

note that sed syntax differs between BSD and GNU. You'll need to use `sed -i ''` (space between `i` and empty string) on BSD.

(this also has the benefit of not breaking git bisect)

I still think it's a good idea to hide the deprecated functionality behind a command-line argument to make removing it in the next version easier.

achow101 commented on src/rpc/misc.cpp:
```diff
@@ -214,38 +166,24 @@ UniValue validateaddress(const JSONRPCRequest& request)
-                }
+        try {
+            ret.pushKVs(getaddressinfo(request));
+        } catch (UniValue e) {
```
That is swallowed in the event that `-disablewallet` is used; `getaddressinfo` will return `RPC_METHOD_NOT_FOUND` but `validateaddress` should still work and validate addresses.

jnewbery replied:
In that case, I think it's preferable to test for wallet existence before calling `getaddressinfo`. The following should do it:

```
        if (!::vpwallets.empty()) {
            ret.pushKVs(getaddressinfo(request));
        }
```

achow101 replied:
Done

promag reviewed (commented):
New code (and probably moved code) should take into account developer notes.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3169,6 +3170,114 @@ UniValue generate(const JSONRPCRequest& request)
+            "1. \"address\"     (string, required) The bitcoin address to get the information of.\n"
+            "\nResult:\n"
+            "{\n"
+            "  \"address\" : \"address\", (string) The bitcoin address validated\n"
```
Forgot to say that this was addressed.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3169,6 +3170,114 @@ UniValue generate(const JSONRPCRequest& request)
+public:
+    CWallet * const pwallet;
+
+    DescribeAddressVisitor(CWallet *_pwallet) : pwallet(_pwallet) {}
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3169,6 +3170,114 @@ UniValue generate(const JSONRPCRequest& request)
+
+    DescribeAddressVisitor(CWallet *_pwallet) : pwallet(_pwallet) {}
+
+    UniValue operator()(const CNoDestination &dest) const { return UniValue(UniValue::VOBJ); }
```
I don't think that is necessary

achow101 commented:
Rebased

jnewbery commented on src/rpc/misc.cpp:
```diff
@@ -127,37 +78,21 @@ UniValue validateaddress(const JSONRPCRequest& request)
+            ret.pushKVs(getaddressinfo(request));
+        }
+#endif
+        if (ret["address"].isNull()) {
```
nit: indentation for this if block

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,123 @@ UniValue generate(const JSONRPCRequest& request)
+UniValue getaddressinfo(const JSONRPCRequest& request)
+{
+    CWallet * const pwallet = GetWalletForJSONRPCRequest(request);
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp))
```
nit: braces

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,123 @@ UniValue generate(const JSONRPCRequest& request)
+    CScript scriptPubKey = GetScriptForDestination(dest);
+    ret.push_back(Pair("scriptPubKey", HexStr(scriptPubKey.begin(), scriptPubKey.end())));;
+    
+    ret.push_back(Pair("isscript", scriptPubKey.IsPayToScriptHash()));
```
micronit: not in same ordering as help text (either for `getaddressinfo` or `validateaddress`. Suggest you move this line below `ret.push_back(Pair("iswatchonly", ...`

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,123 @@ UniValue generate(const JSONRPCRequest& request)
+            "\nReturn information about the given bitcoin address. Some information requires the address\n"
+            "to be in the wallet.\n"
+            "\nArguments:\n"
+            "1. \"address\"     (string, required) The bitcoin address to get the information of.\n"
```
nit: help text alignment

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,123 @@ UniValue generate(const JSONRPCRequest& request)
+            "1. \"address\"     (string, required) The bitcoin address to get the information of.\n"
+            "\nResult:\n"
+            "{\n"
+            "  \"address\" : \"address\", (string) The bitcoin address validated\n"
```
nit: help text alignment

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,123 @@ UniValue generate(const JSONRPCRequest& request)
+            "  \"sigsrequired\" : xxxxx        (numeric, optional) Number of signatures required to spend multisig output\n"
+            "  \"pubkey\" : \"publickeyhex\",    (string) The hex value of the raw public key\n"
+            "  \"iscompressed\" : true|false,  (boolean) If the address is compressed\n"
+            "  \"timestamp\" : timestamp,        (number, optional) The creation time of the key if available in seconds since epoch (Jan 1 1970 GMT)\n"
```
nit: help text alignment

jnewbery commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,45 @@
             "\nReturn information about the given bitcoin address.\n"
+            "Parts of this command have been deprecated and moved to getaddressinfo\n"
             "\nArguments:\n"
             "1. \"address\"     (string, required) The bitcoin address to validate\n"
```
nit: help text alignment

jnewbery commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,45 @@
             "\nResult:\n"
             "{\n"
             "  \"isvalid\" : true|false,       (boolean) If the address is valid or not. If not, this is the only property returned.\n"
             "  \"address\" : \"address\", (string) The bitcoin address validated\n"
```
nit: help text alignment

jnewbery commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,45 @@
-            "  \"timestamp\" : timestamp,        (number, optional) The creation time of the key if available in seconds since epoch (Jan 1 1970 GMT)\n"
-            "  \"hdkeypath\" : \"keypath\"       (string, optional) The HD keypath if the key is HD and available\n"
-            "  \"hdmasterkeyid\" : \"<hash160>\" (string, optional) The Hash160 of the HD master pubkey\n"
+            "  \"timestamp\" : timestamp,        (number, optional) DEPRECATED. The creation time of the key if available in seconds since epoch (Jan 1 1970 GMT)\n"
```
nit: help text alignment

jnewbery reviewed (commented):
A bunch of nits inline. I'm currently running this over the extended test suite to verify no regressions.

My high-level feedback hasn't been addressed:

> we should hide the deprecated fields in validateaddress behind a command-line argument (similar to #11031).

We need to do something to phase out deprecated RPC methods over a couple of releases. The proposal in #11031 is my suggested way of doing that, but I'm open to other suggestions.

jnewbery commented:
extended test suite passes. Travis is failing the scripted diff check.

achow101 commented:
Addressed most of @jnewbery's comments except for the deprecation thing

jnewbery commented:
Thanks for addressing the nits.

My preference is that this doesn't get merged without some kind of phased deprecation. I'm happy to provide a commit to cherry pick if that would help.

achow101 commented:
> My preference is that this doesn't get merged without some kind of phased deprecation. I'm happy to provide a commit to cherry pick if that would help.

I'm planning on making a generic deprecation interface type thing that I will then rebase this and some other stuff on top of.

achow101 commented:
Rebased onto #11031 and moved the deprecated parts behind `IsDeprecatedRPCEnabled`. I wasn't sure what to do with the help text so I just left it as is (except for info about using `-deprecatedrpc`).

jnewbery commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,46 @@
         throw std::runtime_error(
             "validateaddress \"address\"\n"
             "\nReturn information about the given bitcoin address.\n"
+            "Parts of this command have been deprecated and moved to getaddressinfo. To have those deprecated\n"
```
Suggestion only: Perhaps start this with `DEPRECATION WARNING:` to draw attention. It might be easy to miss this warning.

achow101 replied:
Done

jnewbery reviewed (commented):
Looks good. Tested ACK c288c958e96d9c2c856abcde901518f0de9fd3ce.

One suggestion inline, which you could take or leave.

A test for the difference between the return object when `-deprecatedrpc=validateaddress` is set and not would be a nice to have, but not essential.

achow101 commented:
@jnewbery I added a super basic test to check that `-deprecatedrpc=validateaddress` changes what is returned.

promag commented on src/rpc/misc.cpp:
```diff
@@ -127,37 +79,21 @@ UniValue validateaddress(const JSONRPCRequest& request)
 #endif
+        if (ret["address"].isNull()) {
+            std::string currentAddress = EncodeDestination(dest);
+            ret.push_back(Pair("address", currentAddress));
```
Comply with #11386 here and below?

achow101 replied:
Done.

promag commented on src/wallet/rpcwallet.h:
```diff
@@ -10,6 +10,7 @@
 class CRPCTable;
 class CWallet;
 class JSONRPCRequest;
+class UniValue;
```
Why?

achow101 replied:
`UniValue` is needed below for `UniValue getaddressinfo...`

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,125 @@ UniValue generate(const JSONRPCRequest& request)
+    ret.push_back(Pair("scriptPubKey", HexStr(scriptPubKey.begin(), scriptPubKey.end())));;
+
+    isminetype mine = pwallet ? IsMine(*pwallet, dest) : ISMINE_NO;
+    ret.push_back(Pair("ismine", (mine & ISMINE_SPENDABLE) ? true : false));
```
`bool(mine & ISMINE_SPENDABLE)`? Same below.

jnewbery commented on test/functional/deprecated_rpc.py:
```diff
@@ -0,0 +1,29 @@
+
+        # test validateaddress deprecation
+        dep_validate_address = self.nodes[0].validateaddress("mnvGjUy3NMj67yJ6gkK5o9e5RS33Z2Vqcu")
+        assert("ismine" not in dep_validate_address)
```
assert is a statement, not a function:

```python
assert "ismine" not in dep_validate_address
```

same for assert below.

achow101 replied:
Done

jnewbery commented on test/functional/deprecated_rpc.py:
```diff
@@ -0,0 +1,29 @@
+        self.log.info("Using -deprecatedrpc=estimatefee bypasses the error")
+        self.nodes[1].estimatefee(1)
+
+        # test validateaddress deprecation
```
suggestion: change this to a `self.log.info()`

achow101 replied:
Done

jnewbery commented on test/functional/deprecated_rpc.py:
```diff
@@ -0,0 +1,29 @@
+        self.nodes[1].estimatefee(1)
+
+        # test validateaddress deprecation
+        dep_validate_address = self.nodes[0].validateaddress("mnvGjUy3NMj67yJ6gkK5o9e5RS33Z2Vqcu")
```
Perhaps declare this address as a variable above (with comment) to avoid having magic strings in the RPC method call.

achow101 replied:
Done

promag commented on test/functional/deprecated_rpc.py:
```diff
@@ -19,5 +19,11 @@ def run_test(self):
 
+        # test validateaddress deprecation
+        dep_validate_address = self.nodes[0].validateaddress("mnvGjUy3NMj67yJ6gkK5o9e5RS33Z2Vqcu")
+        assert("ismine" not in dep_validate_address)
```
Is it enough to check `ismine` key? Maybe merge `nodes[0].validateaddress` with `nodes[0].getaddressinfo` and that should contain `nodes[1].validateaddress`?

achow101 replied:
It's enough to determine that they are returning different things as one should have `ismine` and the other not.

jnewbery reviewed (commented):
The changed commit broke the scripted-diff checker. You'll need to add your `-deprecatedrpc=validateaddress` test as a separate commit.

promag reviewed (commented):
Only reviewed *[RPC] Split part of validateaddress into getaddressinfo* commit.

achow101 commented:
> The changed commit broke the scripted-diff checker.

I noticed. I believe that should be fixed now as I changed the script to exclude `deprecatedrpc.py`.

achow101 commented:
Rebased onto master

jnewbery commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,46 @@
             "validateaddress \"address\"\n"
             "\nReturn information about the given bitcoin address.\n"
+            "DEPRECATION WARNING: Parts of this command have been deprecated and moved to getaddressinfo. To have those deprecated\n"
+            "information shown, start bitcoind with -deprecatedrpc=validateaddress"
```
I think this warning should also say something like "Clients must transition to using getaddressinfo to access this information before upgrading to v0.17"

achow101 replied:
Done

jnewbery reviewed (commented):
Looks generally good. One question. If I pass an invalid address to `getaddressinfo`, I get the following output:

```
→ bitcoin-cli getaddressinfo fAkEAdDrEsS
{
  "address": "",
  "scriptPubKey": "",
  "ismine": 0,
  "iswatchonly": 0,
  "isscript": 0
}
```

I think it would be more intuitive to either:
- have an `isvalid` field that returns False (like for `validateaddress`); or
- throw a `RPC_INVALID_ADDRESS_OR_KEY` error

What do you think?

jtimon commented:
Concept ACK

> throw a RPC_INVALID_ADDRESS_OR_KEY error

I like this option.

achow101 commented:
> throw a RPC_INVALID_ADDRESS_OR_KEY error

Did that. I didn't realize that I hadn't added that error there.

jnewbery commented:
ACK ea088845fb8dd903f681c4c08b6bfe412cecc536

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,131 @@ UniValue generate(const JSONRPCRequest& request)
+
+    // Make sure the destination is valid
+    if (!IsValidDestination(dest)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid address");
```
Missing test.

achow101 replied:
Added a brief test for `getaddressinfo` in `wallet.py`

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,131 @@ UniValue generate(const JSONRPCRequest& request)
+    CScript scriptPubKey = GetScriptForDestination(dest);
+    ret.pushKV("scriptPubKey", HexStr(scriptPubKey.begin(), scriptPubKey.end()));
+
+    isminetype mine = pwallet ? IsMine(*pwallet, dest) : ISMINE_NO;
```
`pwallet` is valid because of `EnsureWalletIsAvailable` above.

achow101 replied:
Removed the `pwallet` check.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,131 @@ UniValue generate(const JSONRPCRequest& request)
+{
+    CWallet * const pwallet = GetWalletForJSONRPCRequest(request);
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+            return NullUniValue;
```
Nit, fix indentation.

achow101 replied:
Done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,131 @@ UniValue generate(const JSONRPCRequest& request)
+                a.push_back(EncodeDestination(addr));
+            }
+            obj.push_back(Pair("addresses", a));
+            if (whichType == TX_MULTISIG)
```
Nit, add `{` or single line.

achow101 replied:
Done

TheBlueMatt commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,47 @@
-
-    LOCK2(cs_main, pwallet ? &pwallet->cs_wallet : nullptr);
-#else
     LOCK(cs_main);
```
I do not believe you need cs_main here anymore.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,131 @@ UniValue generate(const JSONRPCRequest& request)
+        );
+    }
+
+    LOCK2(cs_main, pwallet->cs_wallet);
```
I believe you only need cs_wallet here.

achow101 replied:
Done.

TheBlueMatt commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,47 @@
-            "  \"script\" : \"type\"             (string, optional) The output script type. Possible types: nonstandard, pubkey, pubkeyhash, scripthash, multisig, nulldata, witness_v0_keyhash, witness_v0_scripthash\n"
-            "  \"hex\" : \"hex\",                (string, optional) The redeemscript for the p2sh address\n"
-            "  \"addresses\"                   (string, optional) Array of addresses associated with the known redeemscript\n"
+            "  \"ismine\" : true|false,        (boolean) DEPRECATED. If the address is yours or not\n"
```
Why change the ordering here? UniValue returns things ordered in insertion-order (and hence bitcoin-cli and some JSON libraries will print it in that order) so might as well just leave it...having deprecation warnings interspersed is fine, they'll all go away soon enough :)

achow101 replied:
The idea was to group the deprecated things together.

TheBlueMatt commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,47 @@
+            "  \"sigsrequired\" : xxxxx        (numeric, optional) DEPRECATED. Number of signatures required to spend multisig output\n"
+            "  \"pubkey\" : \"publickeyhex\",    (string) DEPRECATED. The hex value of the raw public key\n"
+            "  \"iscompressed\" : true|false,  (boolean) DEPRECATED. If the address is compressed\n"
             "  \"account\" : \"account\"         (string) DEPRECATED. The account associated with the address, \"\" is the default account\n"
```
You removed this from the output, but not the docs. For now I think we should leave the account display here, primarily just because we will want to keep it identical but just rename it label when we do accounts->label conversion.

achow101 replied:
Added the accounts field back to the output.

TheBlueMatt reviewed (commented):
Thanks for tackling one of the ifdef WALLETs!

achow101 commented on src/rpc/misc.cpp:
```diff
@@ -31,94 +31,47 @@
             "\nReturn information about the given bitcoin address.\n"
+            "DEPRECATION WARNING: Parts of this command have been deprecated and moved to getaddressinfo. Clients must\n"
+            "transition to using getaddressinfo to access this information before upgrading to v0.17. To have those deprecated\n"
+            "information shown, start bitcoind with -deprecatedrpc=validateaddress"
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3180,6 +3181,131 @@ UniValue generate(const JSONRPCRequest& request)
+            "  \"ismine\" : true|false,        (boolean) If the address is yours or not\n"
+            "  \"iswatchonly\" : true|false,   (boolean) If the address is watchonly\n"
+            "  \"isscript\" : true|false,      (boolean) If the key is a script\n"
+            "  \"script\" : \"type\"             (string, optional) The output script type. Possible types: nonstandard, pubkey, pubkeyhash, scripthash, multisig, nulldata, witness_v0_keyhash, witness_v0_scripthash\n"
```
I changed that field to `redeemscript`. I didn't make it a separate object.

achow101 commented:
Addressed @TheBlueMatt's and @promag's comments. I also changed two of the `pushKV`s back to `push_back(Pair(...))` because they were not properly showing booleans (they were displayed as 0 and 1).

MarcoFalke commented:
Needs rebase

MarcoFalke commented on test/functional/wallet.py:
```diff
@@ -426,6 +426,15 @@ def run_test(self):
 
         # Verify nothing new in wallet
         assert_equal(total_txs, len(self.nodes[0].listtransactions("*",99999)))
+        
```
No trailing whitespace, please.

achow101 replied:
Removed.

achow101 commented:
Rebased onto master.

achow101 commented:
n'th rebase

MeshCollider commented on src/rpc/misc.cpp:
```diff
@@ -32,176 +32,71 @@
+            "  \"iswitness\" : true|false,     (boolean) DEPRECATED. If the address is a witness address\n"
+            "  \"witness_version\" : version   (numeric, optional) DEPRECATED. The version number of the witness program\n"
+            "  \"witness_program\" : \"hex\"   (string, optional) DEPRECATED. The hex value of the witness program\n"
+            "  \"isscript\" : true|false,      (boolean) DEPRECATED. If the key is a script\n"
```
Is this deprecated or not? `isscript` is on line 51 too which hasn't been deleted, and in the if-statement below `Pair("isscript", scriptPubKey.IsPayToScriptHash())` is pushed on line 98. I think the deprecation is correct, so both those lines should be removed?

achow101 replied:
I think that `isscript` shouldn't be deprecated, so I fixed that

MeshCollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3287,6 +3288,179 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+            "    ]\n"
+            "  \"sigsrequired\" : xxxxx        (numeric, optional) Number of signatures required to spend multisig output\n"
+            "  \"pubkey\" : \"publickeyhex\",    (string) The hex value of the raw public key\n"
+            "  \"iscompressed\" : true|false,  (boolean) If the address is compressed\n"
```
Should be `optional`. Same as `pubkey` on line above.
Then make same changes to the deprecated fields in `validateaddress` so they're consistent.

achow101 replied:
Done.

MeshCollider commented on src/rpc/misc.cpp:
```diff
@@ -32,176 +32,71 @@
+            "  \"witness_version\" : version   (numeric, optional) DEPRECATED. The version number of the witness program\n"
+            "  \"witness_program\" : \"hex\"   (string, optional) DEPRECATED. The hex value of the witness program\n"
+            "  \"isscript\" : true|false,      (boolean) DEPRECATED. If the key is a script\n"
+            "  \"redeemscript\" : \"type\"       (string, optional) DEPRECATED. The type of the redeemscript. Possible types: nonstandard, pubkey, pubkeyhash, scripthash, multisig, nulldata, witness_v0_keyhash, witness_v0_scripthash\n"
```
If the deprecation switch is enabled, should `redeemscript` be named `script` for compatibility in the output of `validateaddress`? Otherwise its a breaking change anyway

achow101 replied:
Done

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/10583/commits/aa32c679e75e18996a26f2d4659f91cfa5654964, generally looks good, modulo some feedback inline.

In the case of P2SH-P2WPKH and P2SH-P2WSH, IMO the output should include the `witness_version` and `witness_program` of the internal witness too. Although sipa mentioned he may have done this or something similar in 11403 I think

The deprecation warning mentions v0.17 specifically, perhaps this should have a 0.16.0 milestone, or that may need to be modified

Needs another rebase too sorry :)

achow101 commented:
Rebased and addresses @MeshCollider's comments.

The P2SH nested witness info is being done in #11403.

jnewbery committed:
[rpc] split wallet and non-wallet parts of DescribeAddressVisitor

jnewbery committed:
[rpc] Move DescribeAddressVisitor to rpc/util

jnewbery commented:
Sorry for the late feedback after several rebases, but looking at this again I don't really like DescribeAddressVisitor moving into bitcoin-wallet, since there's functionality in there that's useful even when running with wallet disabled.

Can you take a look at https://github.com/jnewbery/bitcoin/commits/pr10583.1 . It's almost the same as this, but I've moved the generic parts of DescribeAddressVisitor into rpc/util, so they can be accessed by bitcoin-server.

achow101 commented:
@jnewbery I'll take a look.

achow101 commented:
I cherry-picked @jnewbery's commit for splitting and moving `DescribeAddressVisitor`.

jnewbery commented:
I think this is better, but obviously needs more review from others.

Incidentally, I think that you've made this more difficult to review by squashing my commits down into a single commit (having a commit that splits a function into two *and* moves half of it to a new file is more difficult to review than one commit to split and a second commit that is move-only). You've also:

- re-added the `DEPRECATED` notice to `iswitness`, `witness_version` and `witness_program` when they're not actually now deprecated
- re-introduced the `script`/`redeemscript` error in the help text.

achow101 commented:
I cherry-picked the split and move commits from @jnewbery and I think I fixed the regressions

They actually weren't squashed earlier. Rather I had only cherry picked one commit due to merge conflict stuff. That is resolved by taking the commits before mine.

achow101 commented:
Rebased this but there's a linker error I can't quite figure out.

MeshCollider commented:
@achow101: I guess it might be because you're adding rpc/util.cpp to libbitcoin_server_a_SOURCES rather than libbitcoin_util_a_SOURCES in the Makefile.am
Also the scripted diff seems to have an issue

achow101 commented:
Rebased and fixed the linker error.

achow101 commented:
Rebased

jamesob commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,55 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+
+UniValue DescribeAddressVisitor::operator()(const CKeyID &keyID) const {
+    UniValue obj(UniValue::VOBJ);
+    CPubKey vchPubKey;
```
In https://github.com/bitcoin/bitcoin/pull/10583/commits/4dadd7c7ac0934447b8ca33942535cc8a22e78f5

Unused variable?

jamesob commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,55 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+
+UniValue DescribeAddressVisitor::operator()(const CScriptID &scriptID) const {
+    UniValue obj(UniValue::VOBJ);
+    CScript subscript;
```
In https://github.com/bitcoin/bitcoin/pull/10583/commits/4dadd7c7ac0934447b8ca33942535cc8a22e78f5

Unused?

jamesob commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,55 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+UniValue DescribeAddressVisitor::operator()(const WitnessV0KeyHash& id) const
+{
+    UniValue obj(UniValue::VOBJ);
+    CPubKey pubkey;
```
In https://github.com/bitcoin/bitcoin/pull/10583/commits/4dadd7c7ac0934447b8ca33942535cc8a22e78f5

Unused?

jamesob commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,55 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+UniValue DescribeAddressVisitor::operator()(const WitnessV0ScriptHash& id) const
+{
+    UniValue obj(UniValue::VOBJ);
+    CScript subscript;
```
In https://github.com/bitcoin/bitcoin/pull/10583/commits/4dadd7c7ac0934447b8ca33942535cc8a22e78f5

Unused?

jamesob commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,55 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+UniValue DescribeAddressVisitor::operator()(const WitnessUnknown& id) const
+{
+    UniValue obj(UniValue::VOBJ);
+    CScript subscript;
```
In https://github.com/bitcoin/bitcoin/pull/10583/commits/4dadd7c7ac0934447b8ca33942535cc8a22e78f5

Unused?

jamesob commented on src/rpc/util.h:
```diff
@@ -16,4 +23,17 @@ CPubKey HexToPubKey(const std::string& hex_in);
 CPubKey AddrToPubKey(CKeyStore* const keystore, const std::string& addr_in);
 CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey>& pubkeys);
 
+class DescribeAddressVisitor : public boost::static_visitor<UniValue>
```
In https://github.com/bitcoin/bitcoin/pull/10583/commits/4dadd7c7ac0934447b8ca33942535cc8a22e78f5

Curious: if we're looking to eventually get away from using Boost, should we reserve continued/new usage of it for cases when the feature at hand would be hard to replicate on our own? Seems like each use of
```
boost::apply_visitor(DescribeAddressVisitor(), foo);
```
could be replaced with `DescribeAddressVisitor()(foo)`. 

I guess usages of `static_visitor` are relatively easy to unwind if we ever want to, and it may be worthwhile to remain consistent with our usage of it in the wallet code.

Feel free to ignore; not proposing we hold this changeset up on the basis of Boost usage, just curious.

jamesob reviewed (commented):
Tested concept ACK https://github.com/bitcoin/bitcoin/pull/10583/commits/1de28c236c593fc3af770e901ae5d2977370be3f

A few small comments attached. I think the only blocking ones are about unused variables.

achow101 commented:
Removed the unused variables.

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/pull/10583/commits/8a82bf5dc6355486e1b5a86346812c33bc454a1d

Nice work (and a heroic amount of rebasing).

MeshCollider commented:
Hopefully this can go in soon now 0.16 is branched off, looks like a painful one to keep rebasing. Will re-review as soon as I can

achow101 commented:
Rebased yet again.

instagibbs reviewed (approved):
utACK

unused variables don't break anything and quickly disappear, just noting them

sipa commented:
Looking very good, and the commits are easy to follow. A few comments:
* The unused variables @instagibbs pointed out are removed after your move commit, but not before.
* I think `addresses` and `pubkeys` also moved to `getaddressinfo`, but aren't listed in the help.
* Do we need to keep the full description of all deprecated fields in `validateaddress`? I think you could just say at the top "DEPRECATION WARNING: The following fields have moved to getaddressinfo and are only included here when running with -deprecatedrpc=validateaddress: ismine, iswatchonly, script, sigsrequired, pubkey, embedded, iscompressed, timestamp, hdkeypath, hdmasterkeyid".
* Can we remove `addresses` from `getaddressinfo` at the same time? It could remain in `validateaddress` when `-deprecatedrpc=validateaddress` is given, but this is the perfect opportunity to remove it in favor of `pubkeys` for new users.

achow101 commented:
- Removed the extra unused variables
 - Shortened `validateaddress` help
 - Set `addresses` to only be shown if `-deprecatedrpc=validateaddress` is set.

sipa commented on src/rpc/misc.cpp:
```diff
@@ -33,221 +33,56 @@
+            "DEPRECATION WARNING: Parts of this command have been deprecated and moved to getaddressinfo. Clients must\n"
+            "transition to using getaddressinfo to access this information before upgrading to v0.18. The following deprecated\n"
+            "fields have moved to getaddressinfo and will only be shown here with -deprecatedrpc=validateaddress: ismine, iswatchonly,\n"
+            "script, hex, pubkeys, sigsrequired, pubkey, embedded, iscompressed, account, timestamp, hdkeypath, kdmasterkeyid.\n"
```
There probably still needs to be a mention of `addresses`, as there is now no information about it anywhere anymore.

achow101 replied:
Added mention of it back.

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3515,6 +3516,200 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+        // For that reason, only add the 'addresses' field when needed for backward compatibility. New applications
+        // can use the 'embedded'->'address' field for P2SH or P2WSH wrapped addresses, and 'pubkeys' for
+        // inspecting multisig participants.
+        if (IsDeprecatedRPCEnabled("validateaddress")) obj.pushKV("addresses", std::move(a));
```
This will include the `addresses` field when `-deprecatedrpc=validateaddress` is set, even for witness addresses (currently it will only include it for legacy addresses).

achow101 replied:
I changed it back to the original one and used `IsDeprecatedRPC` in a different place to handle this issue.

sipa commented on src/rpc/util.h:
```diff
@@ -16,4 +23,17 @@ CPubKey HexToPubKey(const std::string& hex_in);
 CPubKey AddrToPubKey(CKeyStore* const keystore, const std::string& addr_in);
 CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey>& pubkeys);
 
+class DescribeAddressVisitor : public boost::static_visitor<UniValue>
```
I think you can keep the `DescribeAddressVisitor` class entirely local to `rpc/util.cpp`, and only expose a `UniValue DescribeAddress(const CTxDestination& dest)` function.

achow101 replied:
Changed this and `DescribeWalletAddressVisitor` to helper functions which called the visitors.

achow101 committed:
scripted-diff: validateaddress to getaddressinfo in tests

Change all instances of validateaddress to getaddressinfo since it seems that
no test actually uses validateaddress for actually validating addresses.

-BEGIN VERIFY SCRIPT-
find ./test/functional -path '*py' -not -path ./test/functional/wallet_disable.py -not -path ./test/functional/rpc_deprecated.py -not -path ./test/functional/wallet_address_types.py -exec sed -i'' -e 's/validateaddress/getaddressinfo/g' {} \;
-END VERIFY SCRIPT-

promag commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,61 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+UniValue DescribeAddress(CTxDestination dest)
+{
+    return boost::apply_visitor(DescribeAddressVisitor(), dest);
+}
```
Nit, add newline.

achow101 replied:
Done

promag commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,61 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+        return obj;
+    }
+
+    UniValue operator()(const CScriptID &scriptID) const {
```
Nit, `const CScriptID& scriptID`. `{` on new line.

achow101 replied:
Done

promag commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,61 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+public:
+    explicit DescribeAddressVisitor() {}
+
+    UniValue operator()(const CNoDestination &dest) const {
```
Nit, `const CNoDestination& dest `. `{` on new line.

achow101 replied:
Done

promag commented on src/rpc/misc.cpp:
```diff
@@ -33,221 +33,56 @@
+            "  \"isscript\" : true|false,      (boolean) If the key is a script\n"
+            "  \"iswitness\" : true|false,     (boolean) If the address is a witness address\n"
+            "  \"witness_version\" : version   (numeric, optional) The version number of the witness program\n"
+            "  \"witness_program\" : \"hex\"   (string, optional) The hex value of the witness program\n"
```
Fix alignment.

achow101 replied:
Done

promag commented on src/Makefile.am:
```diff
@@ -335,6 +334,7 @@ libbitcoin_common_a_SOURCES = \
   scheduler.cpp \
   script/sign.cpp \
   script/standard.cpp \
+  script/ismine.cpp \
```
Nit, sort.

achow101 replied:
Done

promag commented on src/Makefile.am:
```diff
@@ -389,10 +389,10 @@ endif
 
 bitcoind_LDADD = \
   $(LIBBITCOIN_SERVER) \
+  $(LIBBITCOIN_WALLET) \
```
Why this change? Does the order matter? 🙄

achow101 replied:
Yes, the order matters

promag commented on src/rpc/util.h:
```diff
@@ -16,4 +23,6 @@ CPubKey HexToPubKey(const std::string& hex_in);
 CPubKey AddrToPubKey(CKeyStore* const keystore, const std::string& addr_in);
 CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey>& pubkeys);
 
+UniValue DescribeAddress(CTxDestination dest);
```
`const CTxDestination& dest`.

achow101 replied:
Done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -27,6 +27,7 @@
 #include <wallet/feebumper.h>
 #include <wallet/wallet.h>
 #include <wallet/walletdb.h>
+#include <wallet/rpcwallet.h>
```
Nit, sort.

achow101 replied:
Done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3515,6 +3516,207 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+class DescribeWalletAddressVisitor : public boost::static_visitor<UniValue>
+{
+public:
+    CWallet * const pwallet;
```
Nit, m_wallet, why public?

achow101 replied:
why not?

promag replied:
See `CAffectedKeysVisitor`, `DestinationEncoder` and `CScriptVisitor` with private members. On the other hand `Witnessifier` members are public (could be "fixed") .

IMO all members, by default, should be private. When reading the code it's clear that a private member is not externally accessed/changed.

promag reviewed (commented):
Needs rebase, some comments (mostly nits).

promag commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,61 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+        return UniValue(UniValue::VOBJ);
+    }
+
+    UniValue operator()(const CKeyID &keyID) const {
```
Nit, `const CKeyID& keyID`. `{` on new line.

achow101 replied:
Done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3514,6 +3515,207 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+
+UniValue DescribeWalletAddress(CWallet* pwallet, const CTxDestination& dest)
+{
+    UniValue ret(UniValue::VOBJ);
```
Nit, could `AssertLockHeld(pwallet->cs_wallet)`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3514,6 +3515,207 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+    UniValue operator()(const CScriptID &scriptID) const {
+        UniValue obj(UniValue::VOBJ);
+        CScript subscript;
+        if (pwallet && pwallet->GetCScript(scriptID, subscript)) {
```
In this context `pwallet` is valid so:
```cpp
if (pwallet->GetCScript(scriptID, subscript)) {
```

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3514,6 +3515,207 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+    UniValue operator()(const CKeyID &keyID) const {
+        UniValue obj(UniValue::VOBJ);
+        CPubKey vchPubKey;
+        if (pwallet && pwallet->GetPubKey(keyID, vchPubKey)) {
```
In this context `pwallet` is valid so:
```cpp
if (pwallet->GetCScript(scriptID, subscript)) {
```

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3514,6 +3515,207 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+        CScript subscript;
+        CRIPEMD160 hasher;
+        uint160 hash;
+        hasher.Write(id.begin(), 32).Finalize(hash.begin());
```
Unless you don't want to touch moved code, remove `hasher` and just do:
```cpp
CRIPEMD160().Write(id.begin(), 32).Finalize(hash.begin());
```

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3514,6 +3515,207 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+        CRIPEMD160 hasher;
+        uint160 hash;
+        hasher.Write(id.begin(), 32).Finalize(hash.begin());
+        if (pwallet && pwallet->GetCScript(CScriptID(hash), subscript)) {
```
In this context `pwallet` is valid so:
```cpp
if (pwallet->GetCScript(CScriptID(hash), subscript)) {
```

promag commented on src/rpc/misc.cpp:
```diff
@@ -33,221 +33,56 @@
-                    ret.pushKV("hdmasterkeyid", meta->hdMasterKeyID.GetHex());
-                }
-            }
+        if (!::vpwallets.empty() && IsDeprecatedRPCEnabled("validateaddress")) {
```
Why `!::vpwallets.empty()`? There is alwasy one wallet right?

achow101 replied:
No, you can run with `-disablewallet` which will disable wallets and thus `vpwallets` will be empty.

promag commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,64 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+        return UniValue(UniValue::VOBJ);
+    }
+
+    UniValue operator()(const CKeyID &keyID) const
```
`&` before space.

promag commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,64 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+        return obj;
+    }
+
+    UniValue operator()(const CScriptID &scriptID) const
```
`&` before space.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3514,6 +3515,207 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+    {
+        UniValue obj(UniValue::VOBJ);
+        CPubKey pubkey;
+        if (pwallet && pwallet->GetPubKey(CKeyID(id), pubkey)) {
```
In this context `pwallet` is valid so:
```cpp
if (pwallet->GetPubKey(CKeyID(id), pubkey)) {
```

promag reviewed (commented):
Lightly tested ACK. Please consider the following comments.

jnewbery commented:
@promag - personally, I'd generally discourage nits after a PR has been open for many months and been rebased/re-reviewed many times already. Adding additional churn and workload to the contributor and reviewers at this point seems borderline sadistic :slightly_smiling_face: 

(just my opinion though)

instagibbs commented on src/rpc/util.cpp:
```diff
@@ -66,3 +66,64 @@ CScript CreateMultisigRedeemscript(const int required, const std::vector<CPubKey
+public:
+    explicit DescribeAddressVisitor() {}
+
+    UniValue operator()(const CNoDestination &dest) const 
```
travis linter is complaining about trailing whitespace.... I don't know protocol if it matters but just reporting where the complaint is coming from

achow101 commented:
Addressed some of @promag's nits. I don't really feel like fixing the other ones. It's also getting really tedious to do fix nits at this point, so I probably won't do so in the future for this PR.

jnewbery commented:
Tested ACK b22cce014852b082d80f1cc35f902b375cba0318. Travis failure looks unrelated.

Let's get this merged!

sipa commented:
utACK b22cce014852b082d80f1cc35f902b375cba0318

I have a few minor improvements, but I'll submit them as a separate PR after merge, nothing worth holding this up for.

promag commented:
@jnewbery @achow101 sorry! didn't want to cause such pain :) I did report tested ACK.

jonasschnelli commented:
Tested ACK b22cce014
